### PR TITLE
test: cancelling both noscrub *and* nodeep-scrub

### PR DESCRIPTION
**Pull request by ronen-fr**
as part of osd-scrub-test.sh.

Signed-off-by: Ronen Friedman <rfriedma@redhat.com>

### CODE REVIEW

**Comment by ronen-fr** [comment]
make-check failed for the unrelated https://tracker.ceph.com/issues/45694.

jenkins retest this please

**Comment by dzafman** [comment]
```
../qa/standalone/scrub/osd-scrub-test.sh:377: _scrub_abort:  local last_scrub=2020-12-07T18:04:21.415287+0000
../qa/standalone/scrub/osd-scrub-test.sh:378: _scrub_abort:  ceph config set osd osd_scrub_sleep 0.1
../qa/standalone/scrub/osd-scrub-test.sh:380: _scrub_abort:  ceph osd unset nodeep-scrub
nodeep-scrub is unset
../qa/standalone/scrub/osd-scrub-test.sh:381: _scrub_abort:  ceph osd unset noscrub
noscrub is unset
```

Strange I ran this by itself and it passed:

../qa/run-standalone.sh "osd-scrub-test.sh TEST_deep_scrub_abort"
PASSED

Running all scrub tests and it failed:

../qa/run-standalone.sh scrub
FAILED

**Comment by ronen-fr** [comment]
@dzafman I am not seeing the failures when testing locally. Can you share the failed log?
(I am testing with my scrub implementation)

**Comment by dzafman** [comment]
> @dzafman I am not seeing the failures when testing locally. Can you share the failed log?
> (I am testing with my scrub implementation)

I'm testing with your scrub implementation and this fix.  I'll upload the log.

**Comment by ronen-fr** [comment]
> I'm testing with your scrub implementation and this fix. I'll upload the log.

@dzafman: seems like it's the same old problem: a race in the test between checking for the aborted scrub
and noticing the new scrub. The scenario is this:

- test sets nodeep
- nodeep is aborted
- regular scrub starts
- test marks the stamp, ostensibly before the next scrub (but really not)
- test waits for a scrub (i.e. for a timestamp change) that has already happened.

I will modify the test to include the line I didn;t post this round: setting 'noscrub' even if only testing for 'nodeep'

**Comment by ronen-fr** [comment]
The other option is to capture the timestamp before we even *set* the nodeep.

**Comment by dzafman** [comment]
> The other option is to capture the timestamp before we even _set_ the nodeep.

Theoretically, it should be sufficient to get the stamp and then _unset noscrub_ and _unset nodeep-scrub_ which starts the scrub again and when it finishes he stamp should be advanced.  I saw the stamp advance in the osd.1 log output I uploaded.  Maybe it took too long?

**Comment by ronen-fr** [comment]
> Theoretically, it should be sufficient to get the stamp and then _unset noscrub_ and _unset nodeep-scrub_ which starts the scrub again and when it finishes he stamp should be advanced. I saw the stamp advance in the osd.1 log output I uploaded. Maybe it took too long?
@dzafman,
Marking down the time just before releasing the "nodeep" is too late. Once the deep scrub was aborted, a regular scrub was immediately scheduled and executed.
For example: in the log you've published:
04:10:32...       no-deep is set
04:10:34.568   the scrubber finished aborting
04:10:37.427   a new scrub was initiated
04:14:02.902   the new (regular) scrub is done
04:14:12.588   the test marks the (wrong) laststamp, and unsets nodeep

I will modify the PR to alway set 'noscrub' if setting 'nodeep'.

**Comment by ronen-fr** [comment]
jenkins retest this please

File: qa/standalone/scrub/osd-scrub-test.sh [hunk 1]
```diff
@@ -351,6 +351,7 @@ function _scrub_abort() {
     fi
 
     ceph osd set $stopscrub
+    ceph osd set noscrub
```
Comment (dzafman) [inline qa/standalone/scrub/osd-scrub-test.sh:354]: I wish this was like the code above that does things conditionally:

``` 
     if [ "$type" = "deep_scrub" ];
      then
       ceph osd set noscrub
     fi
```

File: qa/standalone/scrub/osd-scrub-test.sh [hunk 2]
```diff
@@ -375,6 +376,9 @@ function _scrub_abort() {
     fi
 
     local last_scrub=$(get_last_scrub_stamp $pgid)
+    ceph config set osd "osd_scrub_sleep" "0.1" 
+
+    ceph osd unset $stopscrub
```
Comment (dzafman) [inline qa/standalone/scrub/osd-scrub-test.sh:381]: Same thing here.


```
     ceph osd unset $stopscrub
     if [ "$type" = "deep_scrub" ];
      then
       ceph osd unset noscrub
     fi
```

File: qa/standalone/scrub/osd-scrub-test.sh (hunk 1 above)
Comment (ronen-fr) [reply qa/standalone/scrub/osd-scrub-test.sh:354]: OK w me. Will change and push.

**Review by dzafman — APPROVED** [review]

File: qa/standalone/scrub/osd-scrub-test.sh [hunk 3]
```diff
@@ -375,7 +379,13 @@ function _scrub_abort() {
     fi
 
     local last_scrub=$(get_last_scrub_stamp $pgid)
-    ceph osd unset noscrub
+    ceph config set osd "osd_scrub_sleep" "0.1"
```
Comment (dzafman) [inline qa/standalone/scrub/osd-scrub-test.sh:382]: NIT: There is a space at the end of this line

**Pull request merged**